analyzers/src/SonarAnalyzer.Common/Rules/Utilities/UtilityAnalyzerBase.cs

```diff
@@ -24,18 +24,17 @@
 
 namespace SonarAnalyzer.Rules
 {
+    public readonly record struct UtilityAnalyzerParameter(bool IsAnalyzerEnabled, bool IgnoreHeaderComments, bool AnalyzeGeneratedCode, bool AnalyzeTestProjects, string OutPath, bool IsTestProject)
+    {
+        public static readonly UtilityAnalyzerParameter Default = new(IsAnalyzerEnabled: false, IgnoreHeaderComments: false, AnalyzeGeneratedCode: false, AnalyzeTestProjects: true, OutPath: null, IsTestProject: false);
+    }
+
     public abstract class UtilityAnalyzerBase : SonarDiagnosticAnalyzer
     {
         protected static readonly ISet<string> FileExtensionWhitelist = new HashSet<string> { ".cs", ".csx", ".vb" };
         private readonly DiagnosticDescriptor rule;
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
-        protected bool IsAnalyzerEnabled { get; set; }
-        protected bool IgnoreHeaderComments { get; set; }
-        protected virtual bool AnalyzeGeneratedCode { get; set; }
-        protected virtual bool AnalyzeTestProjects => true;
-        protected string OutPath { get; set; }
-        protected bool IsTestProject { get; set; }
         protected override bool EnableConcurrentExecution => false;
```

Comment:
Like with the `OutPath` property, the mutable state is the root cause that prevents ConcurrentExecution (to be fair: ConcurrentExecution would also work without this change, but it is tough to see why that is the case. With this change, it is evident that it works, and it is no longer possible to break it).

The PR moves the state into its own type, which is now captured in `RegisterCompilationStartAction` for the compilation. The new Parameters struct is `readonly` now and passed to all methods that need it to make decisions.